File: src/200. Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.breakIsland(grid, i, j)
                    count += 1
        return count

    def breakIsland(self, grid: List[List[str]], i: int, j: int):
        m = len(grid)
        n = len(grid[0])
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == '0':
            return
        grid[i][j] = '0'
        self.breakIsland(grid, i + 1, j)
        self.breakIsland(grid, i - 1, j)
        self.breakIsland(grid, i, j + 1)
        self.breakIsland(grid, i, j - 1)

File: src/200. Number of Islands.cpp
class Solution {
private:
    void breakIsland(vector<vector<char>>& grid, int i, int j) {
        int m = grid.size(),
            n = grid[0].size();
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') return;
        grid[i][j] = '0';
        breakIsland(grid, i - 1, j);
        breakIsland(grid, i + 1, j);
        breakIsland(grid, i, j - 1);
        breakIsland(grid, i, j + 1);
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(),
            n = grid[0].size(),
            count = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == '1') {
                    count++;
                    breakIsland(grid, i, j);
                }
            }
        }
        return count;
    }
};